### PR TITLE
dist.to_static support pir program

### PR DESCRIPTION
**Pull request by zhiqiu**
### PR types

New features
### PR changes

APIs
### Description

dist.to_static support pir program
Pcard-76459

In the original static auto-parallel process, the dynamic model is converted into serial program without DistTensors first, and the the dist attributes are added to the program.

However, the new static auto-parallel implementation will infer the spmd info during build operation, so the inputs and parameters in Dynamic model which are DistTensors should be converted to static `DistDenseTensorType` natively and eagerly.

This PR handles the inputs which is sharded by dataloader, the parameters will be handled in the next pr.

For example, given the follow DemoNet, call `shard_dataloader` to make 
```
class DemoNet(nn.Layer):
    def __init__(self, mesh):
        super().__init__()
        self._mesh = mesh
        self.linear_0 = nn.Linear(IMAGE_SIZE, IMAGE_SIZE)
        self.linear_1 = nn.Linear(IMAGE_SIZE, CLASS_NUM)
        self.relu = nn.ReLU()
        # shard the weights of this layer
        self.linear_0.weight = dist.shard_tensor(
            self.linear_0.weight,
            self._mesh,
            [Shard(1)],
            stop_gradient=False,
        )
        self.linear_1.weight = dist.shard_tensor(
            self.linear_1.weight,
            self._mesh,
            [Shard(0)],
            stop_gradient=False,
        )

    def forward(self, x):
        out = self.linear_0(x)
        out = self.relu(out)
        out = self.linear_1(out)
        return out

mesh = dist.ProcessMesh([0, 1], dim_names=["x"])
layer = DemoNet(mesh)
opt = paddle.optimizer.SGD(
    learning_rate=0.1, parameters=layer.parameters()
)
loss_fn = nn.MSELoss()
loader = create_data_loader()
dist_loader = dist.shard_dataloader(loader, meshes=[mesh])
dist_model = dist.to_static(layer, dist_loader, loss_fn, opt)
main_program = dist_model._engine._fwd_main_progs["train"]
```

The serial main program is mixed with DenseTensor and DistDenseTensor. 

```
{
    (%0) = "builtin.parameter" () {is_persistable:[true],parameter_name:"parameter_3",stop_gradient:[false]} : () -> builtin.tensor<8xf32>
    (%1) = "builtin.parameter" () {is_persistable:[true],parameter_name:"parameter_2",stop_gradient:[false]} : () -> builtin.tensor<16x8xf32>
    (%2) = "builtin.parameter" () {is_persistable:[true],parameter_name:"parameter_1",stop_gradient:[false]} : () -> builtin.tensor<16xf32>
    (%3) = "builtin.parameter" () {is_persistable:[true],parameter_name:"parameter_0",stop_gradient:[false]} : () -> builtin.tensor<16x16xf32>
    (%4) = "pd_op.data" () {dtype:(pd_op.DataType)float32,name:"input0",place:(pd_op.Place)Place(undefined:0),shape:(pd_op.IntArray)[4,16],stop_gradient:[true]} : () -> pd_dist.tensor<4x16xf32, mesh: {shape: [2], process_ids: [0,1], dim_names: [x]}, dims_mappings: [-1,-1]>
    (%5) = "pd_op.data" () {dtype:(pd_op.DataType)float32,name:"label0",place:(pd_op.Place)Place(undefined:0),shape:(pd_op.IntArray)[4,8],stop_gradient:[true]} : () -> pd_dist.tensor<4x8xf32, mesh: {shape: [2], process_ids: [0,1], dim_names: [x]}, dims_mappings: [-1,-1]>
    (%6) = "pd_op.matmul" (%4, %3) {stop_gradient:[false],transpose_x:false,transpose_y:false} : (pd_dist.tensor<4x16xf32, mesh: {shape: [2], process_ids: [0,1], dim_names: [x]}, dims_mappings: [-1,-1]>, builtin.tensor<16x16xf32>) -> builtin.tensor<4x16xf32>
    (%7) = "pd_op.add" (%6, %2) {stop_gradient:[false]} : (builtin.tensor<4x16xf32>, builtin.tensor<16xf32>) -> builtin.tensor<4x16xf32>
    (%8) = "pd_op.relu" (%7) {stop_gradient:[false]} : (builtin.tensor<4x16xf32>) -> builtin.tensor<4x16xf32>
    (%9) = "pd_op.matmul" (%8, %1) {stop_gradient:[false],transpose_x:false,transpose_y:false} : (builtin.tensor<4x16xf32>, builtin.tensor<16x8xf32>) -> builtin.tensor<4x8xf32>
    (%10) = "pd_op.add" (%9, %0) {stop_gradient:[false]} : (builtin.tensor<4x8xf32>, builtin.tensor<8xf32>) -> builtin.tensor<4x8xf32>
    (%11) = "pd_op.subtract" (%10, %5) {stop_gradient:[false]} : (builtin.tensor<4x8xf32>, pd_dist.tensor<4x8xf32, mesh: {shape: [2], process_ids: [0,1], dim_names: [x]}, dims_mappings: [-1,-1]>) -> builtin.tensor<4x8xf32>
    (%12) = "pd_op.square" (%11) {stop_gradient:[false]} : (builtin.tensor<4x8xf32>) -> builtin.tensor<4x8xf32>
    (%13) = "pd_op.mean" (%12) {axis:(pd_op.IntArray)[],keepdim:false,stop_gradient:[false]} : (builtin.tensor<4x8xf32>) -> builtin.tensor<f32>
}
```
Currently, the inputs of model, `input0` and `label0`, are DistDenseTensor.
TODO: MAKE parameters be DistDenseTensors.

### CODE REVIEW

File: paddle/fluid/pir/dialect/distributed/ir/dist_dialect.cc [hunk 1]
```diff
@@ -37,18 +39,45 @@ void DistDialect::initialize() {
 void DistDialect::PrintType(pir::Type type, std::ostream &os) const {
   if (auto dist_dense_tensor_type = type.dyn_cast<DistDenseTensorType>()) {
     // Todo: Design the dist dense tensor type print format.
-    os << dist_dense_tensor_type.dense_tensor_type();
+    os << type.dialect().name();
```
Comment (winter-wang) [inline paddle/fluid/pir/dialect/distributed/ir/dist_dialect.cc:42]:
```suggestion
    os << name();
```

File: python/paddle/jit/dy2static/function_spec.py [hunk 2]
```diff
@@ -194,6 +194,26 @@ def pir_to_static_inputs_with_spec(self, input_with_spec, main_program):
...
                     )
                     feed_value.stop_gradient = stop_gradient
+
+                    # warp dist tensor
+                    from paddle.distributed.auto_parallel.static.dist_input_spec import (
+                        DistributedInputSpec,
+                    )
+
+                    if isinstance(var_spec, DistributedInputSpec):
+                        dist_dense_tensor_type = paddle.base.libpaddle.pir.create_dist_dense_tensor_type_by_dense_tensor(
```
Comment (JZ-LIANG) [inline python/paddle/jit/dy2static/function_spec.py:204]: to change to use shard_tensor python API in future when this API is adapted for PIR. @hitywt

File: test/auto_parallel/pir/test_to_static_pir_program.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+        loss_fn = nn.MSELoss()
+        loader = create_data_loader()
+        dist_loader = dist.shard_dataloader(loader, meshes=[mesh])
+        dist_model = dist.to_static(layer, dist_loader, loss_fn, opt)
+
+        dist_model.train()
+        main_program = dist_model._engine._fwd_main_progs["train"]
+        for op in main_program.global_block().ops:
+            tensor = op.result(0)
+            if op.name() == 'pd_op.data':
```
Comment (JZ-LIANG) [inline test/auto_parallel/pir/test_to_static_pir_program.py:98]: enable check for  "builtin.parameter" after shard_tensor API is adapted for PIR. @hitywt

File: test/auto_parallel/pir/test_to_static_pir_program.py [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+                self.assertTrue(tensor.is_dist_dense_tensor_type())
+                self.assertEqual(tensor.process_mesh.shape, [2])
+                self.assertEqual(tensor.process_mesh.process_ids, [0, 1])
+                self.assertEqual(tensor.dims_mapping, [-1, -1])
+                self.assertEqual(tensor.partial_dims, set())
+            else:
```
Comment (JZ-LIANG) [inline test/auto_parallel/pir/test_to_static_pir_program.py:104]: enable check for all other forward computation op after build is adapted for disttensor. @winter-wang

File: paddle/fluid/pir/dialect/distributed/ir/dist_dialect.cc (hunk 1 above)
Comment (zhiqiu) [reply paddle/fluid/pir/dialect/distributed/ir/dist_dialect.cc:42]: Mostlyt, its equal, but I wonder if there is a type which is not belong to dist dialect?

File: python/paddle/jit/dy2static/function_spec.py (hunk 2 above)
Comment (zhiqiu) [reply python/paddle/jit/dy2static/function_spec.py:204]: agree

**Review by JZ-LIANG — APPROVED** [review]
LGTM

**Pull request merged**